File: src/file.lua
-- Disk file handling
--
-- Copyright (c) 2009-2012 Free Software Foundation, Inc.
--
-- This file is part of Zee.
--
-- This program is free software; you can redistribute it and/or modify it
-- under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 3, or (at your option)
-- any later version.
--
-- This program is distributed in the hope that it will be useful, but
-- WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

-- FIXME: Reload when file changes on disk

function exist_file (filename)
  if posix.stat (filename) then
    return true
  end
  local _, err = posix.errno ()
  return err ~= posix.ENOENT
end

local function is_regular_file (filename)
  local st = posix.stat (filename)
  return st and st.type == "regular"
end

-- Return nonzero if file exists and can be written.
local function check_writable (filename)
  local ok = posix.euidaccess (filename, "w")
  return ok and ok >= 0
end

-- Write buffer to given file name with given mode.
alien.default.write:types ("ptrdiff_t", "int", "pointer", "size_t")
local function write_file (filename, mode)
  local h = posix.creat (filename, mode)
  if h then
    local s = get_buffer_pre_point (buf)
    local ret = alien.default.write (h, s.buf.buffer:topointer (), #s)
    if ret == #s then
      s = get_buffer_post_point (buf)
      ret = alien.default.write (h, s.buf.buffer:topointer (), #s)
      if ret == #s then
        ret = true
      end
    end

    local ret2 = posix.close (h)
    if ret == true and ret2 ~= 0 then -- writing error takes precedence over closing error
      ret = ret2
    end

    return ret
  end
end

Define ("file-save",
[[
Save buffer in visited file.
]],
  function ()
    local ret = write_file (buf.filename, "rw-rw-rw-")
    if not ret then
      return minibuf_error (string.format ("Error writing `%s'%s", buf.filename,
                                           ret == -1 and ": " .. posix.errno () or ""))
    end

    minibuf_write ("Wrote " .. buf.filename)
    buf.modified = false
    undo_set_unchanged (buf.last_undop)
  end
)

Define ("file-quit",
[[
Quit, unless there are unsaved changes.
]],
  function ()
    if buf.modified then
      return minibuf_error ("Unsaved changes: use `file-save' or `edit-revert'")
    end

    thisflag.quit = true
  end
)

alien.default.read:types ("ptrdiff_t", "int", "pointer", "size_t")
function read_file (filename)
  if exist_file (filename) and not is_regular_file (filename) then
    return minibuf_error (string.format ("File `%s' exists but could not be read", filename))
  end

  buf = buffer_new ()
  buf.filename = filename
  local len = posix.stat (filename, "size")
  local h = posix.open (filename, posix.O_RDONLY)
  local s
  if h and len then
    s = AStr (len)
    local ret = alien.default.read (h, s:topointer (), len)
    if ret == len then
      buf.readonly = not check_writable (filename)
    end
  end
  buf.text = s or AStr ("") -- FIXME: error if reading fails
end

-- Function called on unexpected error or crash (SIGSEGV).
-- Attempts to save modified buffer.
-- If doabort is true, aborts to allow core dump generation;
-- otherwise, exit.
-- FIXME: Reset the terminal?
function editor_exit (doabort)
  io.stderr:write ("Trying to save buffer (if modified)...\r\n")

  if buf and buf.modified then
    local file = buf.filename .. string.upper (PACKAGE) .. "SAVE"
    io.stderr:write (string.format ("Saving %s...\r\n", file))
    write_to_disk (file, "rw-------")
  end

  if doabort then
    posix.abort ()
  else
    posix._exit (2)
  end
end

File: src/file.c
/* Disk file handling
   Copyright (c) 1997-2004 Sandro Sigala.
   Copyright (c) 2003-2007 Reuben Thomas.
   All rights reserved.

   This file is part of Zee.

   Zee is free software; you can redistribute it and/or modify it under
   the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2, or (at your option) any later
   version.

   Zee is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
   for more details.

   You should have received a copy of the GNU General Public License
   along with Zee; see the file COPYING.  If not, write to the Free
   Software Foundation, Fifth Floor, 51 Franklin Street, Boston, MA
   02111-1301, USA.  */

#include "config.h"

#include <errno.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "main.h"
#include "rbacc.h"
#include "term.h"
#include "extern.h"


/*
 * Get HOME directory.
 * If none, or too long, return blank string.
 */
rblist get_home_dir(void)
{
  char *s = getenv("HOME");

  if (s && strlen(s) < PATH_MAX)
    return rblist_from_string(s);
  else
    return rblist_empty;
}

/*
 * Read the file contents into a string.
 * Return quietly if the file can't be read.
 */
rblist file_read(rblist filename)
{
  FILE *fp;
  if ((fp = fopen(rblist_to_string(filename), "r")) == NULL) {
    return NULL;
  } else {
    rbacc rba = rbacc_add_file(rbacc_new(), fp);
    fclose(fp);
    return rbacc_to_rblist(rba);
  }
}

/*
 * Read the file contents into a buffer.
 * Return quietly if the file can't be read.
 */
void file_open(rblist filename)
{
  buffer_new();
  buf->filename = filename;

  rblist rbl;
  if ((rbl = file_read(buf->filename)) != NULL) {
    // Add lines to buffer
    buf->lines = rbl;
  } else if (errno != ENOENT) {
    buf->flags |= BFLAG_READONLY;
  }
}

/*
 * Write the contents of buffer bp to file filename.
 * The filename is passed separately so a name other than buf->filename
 * can be used, e.g. in an emergency by die.
 * Returns true on success, false on failure.
 */
static bool buffer_write(Buffer *bp, rblist filename)
{
  FILE *fp;
  size_t len = rblist_length(buf->lines);

  assert(bp);

  if ((fp = fopen(rblist_to_string(filename), "w")) == NULL) {
    return false;
  }

  // Save all the lines.
  if (fwrite(rblist_to_string(buf->lines), sizeof(char), len, fp) < len) {
    fclose(fp);
    return false;
  }

  return fclose(fp) == 0;
}

DEF(file_save,
"\
Save buffer in visited file.\
")
{
  if (buffer_write(buf, buf->filename) == false) {
    minibuf_error(rblist_to_string(rblist_fmt("%s: %s", buf->filename, strerror(errno))));
  } else {
    minibuf_write(rblist_to_string(rblist_fmt("Wrote `%r'", buf->filename)));
    buf->flags &= ~BFLAG_MODIFIED;

    if (buf->last_undop)
      undo_reset_unmodified(buf->last_undop);
  }
}
END_DEF

DEF(file_quit,
"\
Quit, unless there are unsaved changes.\
")
{
  if (buf->flags & BFLAG_MODIFIED) {
    minibuf_error("Unsaved changes; use `file_save' or `edit_revert'");
    ok = false;
  }

  if (ok)
    thisflag |= FLAG_QUIT;
}
END_DEF

/*
 * Function called on unexpected error or crash (SIGSEGV).
 * Attempts to save buffer if modified.
 */
// FIXME: Reset the terminal?
void die(int exitcode)
{
  static bool already_dying = 0;

  if (already_dying)
    fprintf(stderr, "die() called recursively; aborting.\r\n");
  else if (buf && buf->flags & BFLAG_MODIFIED) {
    rblist rbl = rblist_empty;
    already_dying = true;
    fprintf(stderr, "Trying to save modified buffer...\r\n");
    if (buf->filename)
      rbl = buf->filename;
    rbl = rblist_concat(rbl, rblist_from_string("." PACKAGE "SAVE"));
    fprintf(stderr, "Saving %s...\r\n", rblist_to_string(rbl));
    buffer_write(buf, rbl);
    term_close();
  }
  exit(exitcode);
}
